icechunk/src/session.rs

```diff
@@ -1361,60 +1365,96 @@ impl<'a> FlushProcess<'a> {
         asset_manager: Arc<AssetManager>,
         change_set: &'a ChangeSet,
         parent_id: &'a SnapshotId,
+        config: &'a RepositoryConfig,
     ) -> Self {
         Self {
             asset_manager,
             change_set,
             parent_id,
+            config,
             manifest_refs: Default::default(),
             manifest_files: Default::default(),
         }
     }
 
-    /// Write a manifest for a node that was created in this session
-    /// It doesn't need to look at previous manifests because the node is new
-    async fn write_manifest_for_new_node(
+    async fn write_manifest_from_iterator(
         &mut self,
         node_id: &NodeId,
-        node_path: &Path,
+        chunks: impl Iterator<Item = ChunkInfo>,
+        shards: ManifestShards,
     ) -> SessionResult<()> {
         let mut from = vec![];
         let mut to = vec![];
-        let chunks = stream::iter(
-            self.change_set
-                .new_array_chunk_iterator(node_id, node_path)
-                .map(Ok::<ChunkInfo, Infallible>),
-        );
-        let chunks = aggregate_extents(&mut from, &mut to, chunks, |ci| &ci.coord);
-
-        if let Some(new_manifest) = Manifest::from_stream(chunks).await.unwrap() {
-            let new_manifest = Arc::new(new_manifest);
-            let new_manifest_size =
-                self.asset_manager.write_manifest(Arc::clone(&new_manifest)).await?;
-
-            let file_info =
-                ManifestFileInfo::new(new_manifest.as_ref(), new_manifest_size);
-            self.manifest_files.insert(file_info);
-
-            let new_ref = ManifestRef {
-                object_id: new_manifest.id().clone(),
-                extents: ManifestExtents::new(&from, &to),
-            };
-
-            self.manifest_refs
-                .entry(node_id.clone())
-                .and_modify(|v| v.push(new_ref.clone()))
-                .or_insert_with(|| vec![new_ref]);
+
+        let mut sharded_refs: HashMap<usize, Vec<_>> = HashMap::new();
+        sharded_refs.reserve(shards.len());
+        // TODO: what is a good capacity
+        let ref_capacity = 8_192;
+        sharded_refs.insert(0, Vec::with_capacity(ref_capacity));
+        for chunk in chunks {
+            let shard_index = shards.which(&chunk.coord)?;
+            sharded_refs
+                .entry(shard_index)
+                .or_insert_with(|| Vec::with_capacity(ref_capacity))
+                .push(chunk);
+        }
```

Comment:
I am attempting to convert this method to accept a `impl Stream<Item=SessionResult<ChunkInfo>>` but I don't see how to convert this groupby logic.